src/Npgsql/TypeHandlers/ArrayHandler.cs

```diff
@@ -279,15 +280,23 @@ public ArrayHandler(PostgresType arrayPostgresType, NpgsqlTypeHandler elementHan
         #region Read
 
         internal override async ValueTask<object> ReadAsObject(NpgsqlReadBuffer buf, int len, bool async, FieldDescription? fieldDescription = null)
-            => typeof(TElement).IsValueType
+            => ReadAsNullable(fieldDescription)
                 ? await ElementTypeInfo<TElement>.ReadNullableArrayFunc(this, buf, async)
                 : await ReadArray<TElement>(buf, async);
 
         internal override object ReadAsObject(NpgsqlReadBuffer buf, int len, FieldDescription? fieldDescription = null)
-            => typeof(TElement).IsValueType
+            => ReadAsNullable(fieldDescription)
                 ? ElementTypeInfo<TElement>.ReadNullableArrayFunc(this, buf, false).GetAwaiter().GetResult()
                 : ReadArray<TElement>(buf, false).GetAwaiter().GetResult();
 
+        [MethodImpl(MethodImplOptions.AggressiveInlining)] 
+        static bool ReadAsNullable(FieldDescription? fieldDescription)
+            => typeof(TElement).IsValueType
```

Comment:
Nit: we tend to put logical operators on the same line at the end, not the next line at the beginning (or on its own line).

Reply:
Ok, fixed.
I still left the `!` at the beginning because in my world view it belongs to the parentheses.
I"ll change that too if you want.